api/query_engine.go

```diff
@@ -18,37 +19,52 @@ var (
 	// metric api.request.render.series is the number of points the request will return.
 	reqRenderPointsReturned = stats.NewMeter32("api.request.render.points_returned", false)
 
-	errUnSatisfiable = errors.New("request cannot be satisfied due to lack of available retentions")
+	errUnSatisfiable   = response.NewError(404, "request cannot be satisfied due to lack of available retentions")
+	errMaxPointsPerReq = response.NewError(413, "request exceeds max-points-per-req-hard limit. Reduce the time range or number of targets or ask your admin to increase the limit.")
 )
 
 // alignRequests updates the requests with all details for fetching, making sure all metrics are in the same, optimal interval
 // note: it is assumed that all requests have the same from & to.
 // also takes a "now" value which we compare the TTL against
 func alignRequests(now uint32, reqs []models.Req) ([]models.Req, error) {
-
 	tsRange := (reqs[0].To - reqs[0].From)
 
 	var listIntervals []uint32
 	var seenIntervals = make(map[uint32]struct{})
+	var targets = make(map[string]struct{})
+
+	for i := range reqs {
+		req := &reqs[i]
+		req.Archive = -1
+		targets[req.Target] = struct{}{}
+	}
+	numTargets := uint32(len(targets))
```

Comment:
i think the idea is that there can be multiple requests with the same `req.Target` but in the response we'll only send each target once, so that's the count of unique targets.